### PR TITLE
[Secure Storage] One more PR than supposed to implement secure storage

### PR DESCRIPTION
**Pull request by ghost**
@davidyuk 

Hey Denis,

So, i've ported the important pieces of the code from  #162 to here, since I wanted to reduce the clutter and mess that became of that branch.

Though currently it works until a certain point. I'm getting an "unhandled promise rejection", I can't seem to pinpoint where is the actual cause.

Saving/Getting the key from SecureStorage works, but there are some other things that don't seem to fit.

Let me know if you have some time to look at it.

### CODE REVIEW

**Comment by sqevani** [comment]
@davidyuk 

I want to wrap my head around the specific functionalities that you're trying to achieve with the keystore and derivedkey, let me know if i'm wrong somewhere.

1. When we signup for a wallet we input a password
2. This password gets salted and is used to encrypt private keys and chainCode
3. derivedKey outside the scope number `2` does not have any other responsibiliy
4. Its simply a workflow to encrypt the account key for the sole purpose of protection

Now:

- We are implementing SecureStorage
- SecureStorage will act on behalf of derivedKey case (which means the keys are going to be stored/encrypted based on users phone pin/signature)
- We don't need derivedKey in this case, since we're not going to store sensitive data on localStorage.

Any issue in this logic? @davidyuk

File: src/store/modules/mobile.js [hunk 1]
```diff
@@ -198,6 +231,77 @@ export default {
...
+      const masterKey = await dispatch('getMasterKey')
+      commit('setMasterKey', masterKey)
+      return undefined
+    },
+
+    /**
+     * Lock masterKey with a derived password
+     * @return {Promise<void>}
+     */
+    async createWithPassword ({ commit, dispatch }, { seed, password }) {
```
Comment (davidyuk) [inline src/store/modules/mobile.js]: `create` what? `createKeystore` was clearer

File: src/store/modules/mobile.js [hunk 2]
```diff
@@ -198,6 +231,77 @@ export default {
...
+
+      if (mac.reduce((p, n) => p || n !== 0, false)) {
+        throw new Error('Invalid password')
+      }
+
+      commit('setDerivedKey', derivedKey)
+    },
+
+    /**
+     * TODO: Follow up on this later
```
Comment (davidyuk) [inline src/store/modules/mobile.js]: I won't merge code style changes in this PR.

File: src/main.js [hunk 3]
```diff
@@ -64,8 +64,8 @@ const initialize = function (App) {
...
  */
-document.addEventListener(
-  process.env.IS_CORDOVA ? 'deviceready' : 'onload',
-  initialize.bind(null, App),
-  false
-)
+if (process.env.IS_CORDOVA) {
+  document.addEventListener('deviceready', initialize.bind(null, App), false)
+} else {
+  window.addEventListener('load', initialize.bind(null, App), false)
```
Comment (davidyuk) [inline src/main.js]: why don't just call `initialize` without any `addEventListener` outside of Cordova as it was before?

File: src/store/modules/mobile.js [hunk 4]
```diff
@@ -23,7 +23,9 @@ const derivePasswordKey = async (password, salt) => {
 
 export default {
   state: {
-    keystore: false,
+    masterKey: null,
+    hasMasterKey: false,
+    keystore: null,
```
Comment (davidyuk) [inline src/store/modules/mobile.js]: `hasMasterKey` is derived from `masterKey`? `masterKey` is the same as `keystore` but in Cordova? Don't make state bigger without necessity.

File: src/store/modules/mobile.js [hunk 5]
```diff
@@ -34,11 +36,24 @@ export default {
...
-    addresses: ({ accounts }) => Object.keys(accounts),
-    loggedIn: ({ keystore, derivedKey }) => !!(keystore && derivedKey)
+    addresses ({ accounts }) {
+      return Object.keys(accounts)
+    },
+    loggedIn ({ masterKey, keystore, derivedKey }) {
+      if (derivedKey) {
+        return !!(keystore && derivedKey)
+      }
+      return !!masterKey
```
Comment (davidyuk) [inline src/store/modules/mobile.js]: Actually, it is just `!!(keystore && derivedKey || masterKey)`

File: src/store/plugins/decryptAccounts.js [hunk 6]
```diff
@@ -3,9 +3,12 @@ import AES from '../../lib/aes'
...
   store.watch(
-    ({ mobile: { keystore, accountCount, derivedKey } }) => [keystore, accountCount, derivedKey],
-    async ([keystore, accountCount, derivedKey]) => {
-      if (!keystore || !derivedKey) return
+    ({ mobile: { masterKey, keystore, accountCount, derivedKey } }) => [masterKey, keystore, accountCount, derivedKey],
+    async ([masterKey, keystore, accountCount, derivedKey]) => {
+      if (!keystore || !derivedKey) {
+        store.commit('setAccounts', getHDWalletAccounts(masterKey, accountCount))
+        return
+      }
```
Comment (davidyuk) [inline src/store/plugins/decryptAccounts.js]: Actually, `decryptAccounts` plugin is useful only outside of Cordova.

**Comment by j28** [comment]
@davidyuk what is the current status of this?

**Comment by etharner** [comment]
Too outdated for current architecture, i'm closing it.

**Pull request closed without merging**